File: Data Structures & Algorithms/is-palindrome/submission-0.cpp
class Solution {
private:
    char lowerCaseChar(char c){
        if( c>= 'A' && c<='Z'){
            return c - 'A' + 'a';
        }

        if( c>= 'a' && c<='z'){
            return c;
        }
    }




    bool isAlphaNumeric(char ch){
        return (ch >= 'A' && ch <= 'Z') || (ch >= 'a' && ch <= 'z') ||
         (ch >= '0' && ch <= '9');
    }
public:
    bool isPalindrome(string s) {
        
        int start = 0;
        int end = s.length() - 1;

        while(start<end){
            
            while( start < end && !isAlphaNumeric(s[start])){
                start++;
            }
             while(start < end && !isAlphaNumeric(s[end])){
                end--;
            }

            if(lowerCaseChar(s[start]) != lowerCaseChar(s[end])){
                return false;
            }else{
                start++;
                end--;
            }

        } 
        return true;
    }
};

File: Data Structures & Algorithms/palindrome-partitioning/submission-0.js
class Solution {
    /**
     * @param {string} s
     * @return {string[][]}
     */
    isPalindrome(str, start, end) {
        while (start <= end) {
            if (str[start++] !== str[end--]) {
                return false;
            }
        }
        return true;
    }

    findPartition(idx, s, path, res) {
        if (idx === s.length) {
            res.push([...path]);
            return;
        }
        for (let i = idx; i < s.length; i++) {
            if (this.isPalindrome(s, idx, i)) {
                path.push(s.substring(idx, i + 1));
                this.findPartition(i + 1, s, path, res);
                path.pop();
            }
        }
    }
    partition(s) {
        let res = [];
        let path = [];
        this.findPartition(0, s, path, res);
        return res;
    }
}
